hooks/emitter/event_querier.go

```diff
@@ -0,0 +1,35 @@
+package emitter
+
+import abci "github.com/cometbft/cometbft/abci/types"
+
+// EventQuerier is a helper struct that helps to find the event with some conditions.
+type EventQuerier struct {
+	events []abci.Event
+}
+
+// NewEventQuerier creates a new EventQuerier with the given events.
+func NewEventQuerier(events []abci.Event) *EventQuerier {
+	return &EventQuerier{events: events}
+}
+
+// FindEventWithTypeBeforeIdx returns the last event of the given type that is emitted before the given index.
+func (es *EventQuerier) FindEventWithTypeBeforeIdx(eventType string, idx int) (abci.Event, bool) {
```

Comment:
TBD: should we limit this or we can search until zero

Reply:
IMO, I think it serves the purpose of this function; cannot know where the matched event type is unless we go through the list.

Reply:
got it!